### PR TITLE
feat(gapic): add OpenTelemetry channel tracing to generator templates

### PR DESCRIPTION
**Pull request by chalmerlowe**
## Problem
Generated client libraries currently lack OpenTelemetry tracing interceptor support for gRPC channels. When OpenTelemetry is configured in client options, generated clients cannot automatically create and pass interceptors down to their transport and gRPC channel.

## Solution
Updates the code in the following ways:
- Imports the centralized `ClientInterceptor` in the gRPC transport template and updates the transport initializer signature and docstrings to accept channel interceptors.
- Updates the client template to look up OpenTelemetry interceptors via `google.api_core._observability` and pass them into the gRPC transport arguments.
- Adds showcase-unit tests to the generated unit test templates to verify interceptor injection and transport application.
- Adds showcase (system tests) to the generated system test templates.

## Non-Goals & Future Work
- **AsyncIO Transport (`grpc_asyncio`)**: Channel tracing for `grpc_asyncio` and `AsyncClient` is intentionally deferred to an upcoming follow-up PR to maintain a focused review scope and ensure isolated test coverage.
- **Streaming RPC Assertions**: While gRPC client interceptors intercept streaming calls, dedicated verification tests for client-, server-, and bidirectional-streaming spans will be added in a subsequent observability milestone.

## Notes for Reviewers
- The changes here are based on the `feat/otel-tracing-t4-resource-attributes` branch which introduces the required core observability helpers.

### CODE REVIEW

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2 [hunk 1]
```diff
@@ -10,7 +10,20 @@ import pickle
...
 from typing import Callable, Dict, Optional, Sequence, Tuple, Union
 
+import grpc  # type: ignore
 from google.api_core import grpc_helpers
+
+# Optional: OpenTelemetry tracing capabilities for grpc channel injection
+# Note: ClientInterceptor was added in google-api-core 2.25.0+; fallback for older versions
+try:
+    from google.api_core.grpc_helpers import ClientInterceptor  # type: ignore[attr-defined]
+except ImportError:  # pragma: NO COVER
```
Comment (chalmerlowe) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:20]: ### Added `do-not-merge`
Until the most recent version of `google-api-core` is published to PyPI, the ClientInterceptor object is unavailable. While working on this PR, this is a temporary workaround to enable testing, etc. Will be removed before merge.

File: packages/google-api-core/google/api_core/_observability.py [hunk 2]
```diff
@@ -64,6 +65,131 @@ def is_otel_capabilities_enabled(
...
+        parsed = urllib.parse.urlsplit(target)
+        if parsed.hostname:
+            attrs["server.address"] = parsed.hostname
+        if parsed.port:
+            scheme = parsed.scheme.lower()
+            is_default_port = (parsed.port == 443 and scheme in ("https", "")) or (
+                parsed.port == 80 and scheme == "http"
+            )
+            if not is_default_port:
+                attrs["server.port"] = parsed.port
```
[the remark this answers is not in the document]
Comment (chalmerlowe) [reply packages/google-api-core/google/api_core/_observability.py]: OBE. Resolved in another PR.

File: packages/google-api-core/google/api_core/gapic_v1/method.py [hunk 3]
```diff
@@ -132,22 +303,67 @@ def __init__(
...
+            and _observability.is_otel_capabilities_enabled(client_options)
+        ):
+            try:
+                from opentelemetry import trace
+
+                tracer_provider = (
+                    getattr(client_options, "tracer_provider", None)
+                    if client_options is not None
+                    else None
+                )
```
[the remark this answers is not in the document]
Comment (chalmerlowe) [reply packages/google-api-core/google/api_core/gapic_v1/method.py]: OBE. Resolved in another PR.

**Comment by chalmerlowe** [comment]
This comment is now OBE.

~~We have not yet re-generated the Goldens, thus these are both failing:~~

~~* `Gapic Generator Specialized Tests / integration`~~
~~* `Gapic Generator Specialized Tests / status` (roll-up of all test status)~~ 

~~The goldens are forth coming.~~

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 4]
```diff
@@ -531,19 +538,39 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
                 else cast(Callable[..., {{ service.name }}Transport], transport)
             )
             {% endif %}
+            # When OpenTelemetry tracing is enabled, obtain the channel interceptor
+            # and pass it to the transport.
+            interceptors = []
+            {% if 'grpc' in opts.transport %}
+            if (
+                isinstance(transport_init, type)
+                and issubclass(transport_init, {{ service.grpc_transport_name }})
```
Comment (chalmerlowe) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:547]: > [!NOTE]
> For reviewers: this line is updated from what was in the prototype.
#### New:
```
    isinstance(transport_init, type)
    and issubclass(transport_init, {{ service.grpc_transport_name }})
```

#### Old (from the `secret-manager` prototype):
```
    transport_init is SecretManagerServiceGrpcTransport
```


There was a concern if we solely checked for "is the object an instance of the given class". If a user subclassed (i.e. `EchoGrpcTransport`) the old check would throw an error. We added two small protections here:

We check for whether the object `issubclass(...)`.
We avoid a `TypeError` during that check by ensuring that we don't pass in a function. These ensure that we only pass interceptors to gRPC transports that are guaranteed to accept the expected parameters.

File: packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/client.py [hunk 5]
```diff
@@ -545,18 +552,36 @@ def __init__(self, *,
                 if isinstance(transport, str) or transport is None
                 else cast(Callable[..., AssetServiceTransport], transport)
             )
+            # When OpenTelemetry tracing is enabled, obtain the channel interceptor
+            # and pass it to the transport.
+            interceptors = []
+            if (
+                isinstance(transport_init, type)
+                and issubclass(transport_init, AssetServiceGrpcTransport)
```
Comment (daniel-sanche) [inline packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/client.py]: Actually looking at this more, it seems like it could make more sense to append these custom interceptors within the transports, instead of passing them down, since each transport requires different types of interceptors

This logic won't add the interceptors if the `transport_init` is a `str` or `Callable`. And the async client [configures its transport with a  `str`](https://github.com/googleapis/google-cloud-python/blob/b409ba6846dde6b2aca225d36e21b3886df1973a/packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/async_client.py#L201), so this logic will get a lot messier soon

Sorry for the back and forth on this. What do you think?

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2 [hunk 6]
```diff
@@ -10,7 +10,20 @@ import pickle
...
+# Note: ClientInterceptor was added in google-api-core 2.36.0+; fallback for older versions
+try:
+    from google.api_core.grpc_helpers import ClientInterceptor  # type: ignore[attr-defined]
+except ImportError:  # pragma: NO COVER
+    ClientInterceptor = Union[  # type: ignore[misc,assignment]
+        grpc.UnaryUnaryClientInterceptor,
+        grpc.UnaryStreamClientInterceptor,
+        grpc.StreamUnaryClientInterceptor,
+        grpc.StreamStreamClientInterceptor,
+    ]
```
Comment (daniel-sanche) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:26]: I know this will be changing soon, but can we put this inside a `TYPE_CHECKING` block, and use a string "ClientInterceptor"?

Even after this is added to api_core, many users won't have the latest version. But we only need this type here when running mypy

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 (hunk 4 above)
Comment (daniel-sanche) [reply packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:547]: This is a good change, but looking at this more, I'm thinking we should probably just append the custom interceptors inside the Transport.\_\_init__, instead of trying to build the interceptor list here. I forgot that this same method is shared for sync/async/rest, which all have different interceptor formats. And this logic loses the interceptors if a `Callable` is passed

What do you think?

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2 [hunk 7]
```diff
@@ -252,6 +275,13 @@ class {{ service.name }}GrpcTransport({{ service.name }}Transport):
                 ],
             )
 
+        apply_interceptors = getattr(
+            grpc_helpers,
+            "apply_channel_interceptors",
+            lambda channel, interceptors: channel,
+        )
```
Comment (daniel-sanche) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:282]: suggestion: If you wanted to make sure this is present, you could make use of the [_compat file](https://github.com/googleapis/google-cloud-python/blob/main/packages/google-cloud-bigtable/google/cloud/bigtable_v2/_compat.py) until we get the right version of api_core in place

Comment (chalmerlowe) [reply packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2:296]: ### Not recommended

Regarding `_compat.py`: Because OpenTelemetry tracing requires `google-api-core >= 2.36.0` anyway (for `_observability` and `method` spans), generated clients on older `api_core` versions will never emit or pass interceptors. Since interceptors is a new parameter that didn't exist in older client releases, the `getattr` fallback `lambda` safely avoids `AttributeError` without needing to generate and maintain a duplicate polyfill in `_compat.py` across 100+ libraries. When `google-api-core >= 2.36.0` is eventually set as a minimum dependency in `setup.py`, we can drop the `getattr` entirely.

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2 (hunk 6 above)
Comment (chalmerlowe) [reply packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/grpc.py.j2]: Done.

**Comment by chalmerlowe** [comment]
> This is a good change, but looking at this more, I'm thinking we should probably just append the custom interceptors inside the Transport.__init__, instead of trying to build the interceptor list here. I forgot that this same method is shared for sync/async/rest, which all have different interceptor formats. And this logic loses the interceptors if a Callable is passed
> 
> What do you think?

I made this change.

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/base.py.j2 [hunk 8]
```diff
@@ -145,17 +158,39 @@ class {{ service.name }}Transport(abc.ABC):
...
+        self._wrap_with_tracing = _WRAP_METHOD_SUPPORTS_TRACING
+
         self._wrapped_methods: Dict[Callable, Callable] = {}
 
     @property
     def host(self):
         return self._host
 
+    def _wrap_method(self, func, *args, **kwargs):
+        if self._wrap_with_tracing:
```
Comment (daniel-sanche) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/base.py.j2:171]: nit: can't this just be `if _WRAP_METHOD_SUPPORTS_TRACING`? It doesn't seem like it should be tied to the transport

File: packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/transports/base.py [hunk 9]
```diff
@@ -122,26 +135,50 @@ def __init__(
...
     def _prep_wrapped_messages(self, client_info):
         # Precompute the wrapped methods.
         self._wrapped_methods = {
-            self.export_assets: gapic_v1.method.wrap_method(
+            self.export_assets: self._wrap_method(
                 self.export_assets,
                 default_timeout=60.0,
                 client_info=client_info,
+                method_name="google.cloud.asset.v1.AssetService/ExportAssets",
             ),
```
Comment (daniel-sanche) [inline packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/transports/base.py:174]: If the point of the wrapper is just to strip out unsupported args, we should be able to avoid the extra indirection, and do something like this:

```
try:
    kind = self.kind
except NotImplementedError:
    kind = None

self.export_assets: gapic_v1.method.wrap_method(
    self.export_assets,
    default_timeout=60.0,
    client_info=client_info,
    **({
           "method_name": "google.cloud.asset.v1.AssetService/ExportAssets",
           "kind": kind,
        } if _WRAP_METHOD_SUPPORTS_TRACING and kind else {}
    )
),
```

File: packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/transports/base.py [hunk 10]
```diff
@@ -122,26 +135,50 @@ def __init__(
...
     @property
     def host(self):
         return self._host
 
+    def _wrap_method(self, func, *args, **kwargs):
+        if self._wrap_with_tracing:
+            kwargs["client_options"] = self._client_options
+            try:
+                kwargs["kind"] = self.kind
+            # The abstract BaseTransport class raises NotImplementedError for the kind property.
```
Comment (daniel-sanche) [inline packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/transports/base.py:152]: I wonder if it would be worth adding a value for the base class, so we don't have to worry about this edge case. Maybe an empty string

File: packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/transports/grpc.py [hunk 11]
```diff
@@ -259,6 +287,22 @@ def __init__(self, *,
                 ],
             )
 
+        channel_interceptors = list(interceptors) if interceptors else []
+        if (
+            _observability is not None
+            and (otel_interceptor := _observability.get_otel_interceptor(self._client_options)) is not None
+            and otel_interceptor not in channel_interceptors
+            and not any(getattr(i, "_is_otel_interceptor", None) is True for i in channel_interceptors)
```
Comment (daniel-sanche) [inline packages/gapic-generator/tests/integration/goldens/asset/google/cloud/asset_v1/services/asset_service/transports/grpc.py:295]: relying on an internal property like this seems a little ugly, since it could change at any time. I'd say this check probably isn't worth it, and we can just trust our inputs

But this is protected by a getattr guard, so not a big deal either way

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 12]
```diff
@@ -11,6 +11,7 @@ from collections import OrderedDict
 import functools
 {% endif %}
 from http import HTTPStatus
+import inspect
```
Comment (daniel-sanche) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:14]: this seems to be unused

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 13]
```diff
@@ -531,19 +539,34 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
+            client_options = None
+            if (
+                _observability is not None
+                and _observability.is_otel_capabilities_enabled(self._client_options)
+                and (
+                    not isinstance(transport_init, type)
+                    or issubclass(transport_init, {{ service.grpc_transport_name }})
+                )
+            ):
+                client_options = self._client_options
```
Comment (daniel-sanche) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2]: Maybe we should pass this in regardless? There may be other use cases for it in the future

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/base.py.j2 [hunk 14]
```diff
@@ -75,6 +84,7 @@ class {{ service.name }}Transport(abc.ABC):
             client_info: gapic_v1.client_info.ClientInfo = DEFAULT_CLIENT_INFO,
             always_use_jwt_access: Optional[bool] = False,
             api_audience: Optional[str] = None,
+            client_options: Optional[Union[client_options_lib.ClientOptions, dict]] = None,
```
Comment (daniel-sanche) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/transports/base.py.j2:87]: Does this need to be added to http too, to be consistent? 

We should at least add a **kwarg, so new arguments can be passed through?

File: packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2 [hunk 15]
```diff
@@ -531,19 +539,34 @@ class {{ service.client_name }}(metaclass={{ service.client_name }}Meta):
...
             {% endif %}
+            # When OpenTelemetry tracing is enabled, pass client_options to the transport
+            # so it can wire tracing interceptors and method spans.
+            client_options = None
+            if (
+                _observability is not None
+                and _observability.is_otel_capabilities_enabled(self._client_options)
+                and (
+                    not isinstance(transport_init, type)
+                    or issubclass(transport_init, {{ service.grpc_transport_name }})
```
Comment (daniel-sanche) [inline packages/gapic-generator/gapic/templates/%namespace/%name_%version/%sub/services/%service/client.py.j2:550]: Do we have to be so selective about which transports we pass client_options to? I was thinking by moving the interceptor to the transport, we could pass the same init args to them all

(I wonder if grpc_transport_name here could cause problems specifically, if there are any http-only clients without this set)

**Review by daniel-sanche — APPROVED** [review]
Looking good for the most part, but take a look at my comments before merging. Particularly around client_options
